src/OpenTelemetry/Resources/ResourceBuilder.cs

```diff
@@ -142,8 +142,7 @@ public ResourceBuilder AddDetector(IResourceDetector resourceDetector)
         /// </remarks>
         /// <param name="resourceDetectorFactory">Resource detector factory.</param>
         /// <returns>Supplied <see cref="ResourceBuilder"/> for call chaining.</returns>
-        // Note: This API may be made public if there is a need for it.
-        internal ResourceBuilder AddDetector(Func<IServiceProvider?, IResourceDetector> resourceDetectorFactory)
+        public ResourceBuilder AddDetector(Func<IServiceProvider?, IResourceDetector> resourceDetectorFactory)
```

Comment:
Spoke with @CodeBlanch offline a bit about this... This method seems kinda wonky in that it differs from our provider builders (tracer, meter, etc). I floated the idea of something like

```csharp
        public static ResourceBuilder AddDetector<T>(this ResourceBuilder resourceBuilder)
            where T : IResourceDetector
        {
            return resourceBuilder.AddDetector(sp => sp.GetRequiredService<T>());
        }
```

Though this is wonky for other reasons. It differs from methods like AddProcessor/Instrumentation/Sampler/etc on our provider builders in that this will not automatically add a detector of type T to the service collection.

Also, @CodeBlanch raised a point that ResourceBuilder is hard because it has a public constructor, so something like the following would never work

```csharp
new ResourceBuilder().AddDetector<MyDetector>()
```

All that said, my hope is we can do something better here that feels more consistent with the provider builders. My vote is to hold off on this change until after 1.4. A [workaround has been provided](https://github.com/open-telemetry/opentelemetry-dotnet/issues/4139#issuecomment-1414229507), albeit not great, but gives us a chance to consider alternatives.

Reply:
Sounds good to me I'll close this and we can work on it for a future release.